qualtran/_infra/data_types.py

```diff
@@ -0,0 +1,189 @@
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Quantum data type definitions.
+
+We often wish to write algorithms which operate on quantum data. One can think
+of quantum data types, similar to classical data types, where a collection of
+qubits can be used to represent a specific quantum data type (eg: a quantum
+integer of width 32 would comprise of 32 qubits, similar to a classical uint32
+type). More generally, many current primitives and algorithms in qualtran
+implicitly expect registers which represent signed or unsigned integers,
+fixed-point (fp) numbers , or “classical registers” which store some classical
+value. Enforcing typing helps developers and users reason about algorithms, and
+will also allow better type checking.
+
+The basic principles we follow are:
+
+1. Typing should not be too invasive for the developer / user: We got pretty far
+without explicitly typing registers.
+2. For algorithms or bloqs which expect registers which are meant to encode
+numeric types (integers, reals, etc.) then typing should be strictly enforced.
+For example, a bloq multiplying two fixed point reals should be built with an
+explicit QFixedPoint dtype.
+3. The smallest addressable unit is a QBit. Other types are interpretations of
+collections of QBits. A QUnsigedInt(32) is intended to represent a register
+encoding positive integers.
+4. To avoid too much overhead we have a QAny type, which is meant to represent
+an opaque bag of bits with no particular significance associated with them. A
+bloq defined with a QAny register (e.g. a n-bit CSwap) will accept any other type assuming the
+bitsizes match. QInt(32) == QAny(32), QInt(32) != QFixedPoint(16, 16). QInt(32) != QUnsignedInt(32).
+5. We assume a big endian convention for addressing QBits in registers throughout qualtran.
+6. Ones' complement integers are used extensively in quantum algorithms. We have
+two types QInt and QIntOnesComp for integers using two's and ones' complement
+respectively.
+"""
+
+import abc
+from typing import Union
+
+import attrs
+import sympy
+
+
+class QDType:
+    @property
+    @abc.abstractmethod
+    def num_qubits(self):
+        """Number of qubits required to represent a single instance of this data type."""
+
+
+@attrs.frozen
+class QBit(QDType):
+    """A single qubit. The smallest addressable unit of quantum data."""
+
+    @property
+    def num_qubits(self):
+        return 1
+
+
+@attrs.frozen
+class QAny(QDType):
+    """Opaque bag-of-qbits type."""
+
+    bitsize: Union[int, sympy.Expr]
+
+    def __attrs_post_init__(self):
+        assert self.bitsize > 1, "bitsize must be > 1."
```

Comment:
this will complain if bitsize is a sympy.Expr

Reply:
Fixed.